### ISSUE TITLE
[BUG] Pydantic Configs are not JSON-serializable

### ISSUE THREAD

**Issue by Quentin-Anthony** [microsoft/DeepSpeed#2441]
@jeffra @mrwyattii @dashstander @tjruwase 

**Describe the bug**
Now that DeepSpeed configs are being migrated to Pydantic, calling `json.dump()` or `json.dumps()` on any config that contains a Pydantic config will fail. For example, with autotuning, we see:

```
Traceback (most recent call last):
  File "/fsx/quentin/autotuning/gpt-neox/deepy.py", line 38, in <module>
    main(deepspeed_main_args)
  File "/fsx/quentin/autotuning/DeepSpeed/deepspeed/launcher/runner.py", line 425, in main
    run_autotuning(args, active_resources)
  File "/fsx/quentin/autotuning/DeepSpeed/deepspeed/launcher/runner.py", line 334, in run_autotuning
    tuner.tune()
  File "/fsx/quentin/autotuning/DeepSpeed/deepspeed/autotuning/autotuner.py", line 435, in tune
    next_max_mbs, next_mbs, next_metric_val = self.tune_space(
  File "/fsx/quentin/autotuning/DeepSpeed/deepspeed/autotuning/autotuner.py", line 541, in tune_space
    min_micro_batch_size, max_micro_batch_size = self.get_min_max_micro_batch_size(
  File "/fsx/quentin/autotuning/DeepSpeed/deepspeed/autotuning/autotuner.py", line 883, in get_min_max_micro_batch_size
    exp, metric_val = self.run_ds_config(ds_config, exp_name)
  File "/fsx/quentin/autotuning/DeepSpeed/deepspeed/autotuning/autotuner.py", line 1082, in run_ds_config
    json.dump(exp_config, fd)
  File "/fsx/quentin/miniconda3/lib/python3.9/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/fsx/quentin/miniconda3/lib/python3.9/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/fsx/quentin/miniconda3/lib/python3.9/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/fsx/quentin/miniconda3/lib/python3.9/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/fsx/quentin/miniconda3/lib/python3.9/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/fsx/quentin/miniconda3/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type DeepSpeedZeroConfig is not JSON serializable
```

which is caused by calling `json.dump()` at line: https://github.com/microsoft/DeepSpeed/blob/480a2212b59081ab79d76a30dbe82fda2be2a82e/deepspeed/autotuning/autotuner.py#L1081 
which fails when it tries to parse the autotuner's Pydantic DeepSpeedZeroConfig here: https://github.com/microsoft/DeepSpeed/blob/480a2212b59081ab79d76a30dbe82fda2be2a82e/deepspeed/autotuning/autotuner.py#L759
This didn't used to be an issue since DeepSpeedZeroConfig used to inherit from DeepSpeedConfigObject, which was introduced in https://github.com/microsoft/DeepSpeed/pull/862/files: https://github.com/microsoft/DeepSpeed/blob/59975896830e8e51c296d072e857a35a71683185/deepspeed/runtime/config_utils.py#L126

**To Reproduce**
1. Enable autotuning (we reproduced on the cifar DeepSpeedExample)

**Proposed Solution**
I was able to resolve this by changing `json.dump(exp_config, fd)` to
```
from pydantic.json import pydantic_encoder
...
json.dump(exp_config, fd, default=pydantic_encoder)
```
which provides json with a function to handle objects that it can't serialize by default (https://docs.python.org/3/library/json.html#json.dump). I propose simply adding the `default=pydantic_encoder` arg to all `json.dump()` and `json.dumps()` calls within DeepSpeed. **If this solution is acceptable to the DeepSpeed team, I can create a PR with this quickly.**

**Comment by mrwyattii** [comment]
@Quentin-Anthony I think this problem has been resolved by @cli99 in #2395. To summarize, the config used in autotuning shouldn't have been converted to a Pydantic config because they are saved to json and ingested by another `deepspeed` process, where the config can be validated with Pydantic.

Also, I'm currently working on moving all of the DeepSpeed configs to using Pydantic in #2210 and this will allow us to directly get a dict or json string to dump to file, like this:
```python
from pydantic import BaseModel
import json

class SubConfig(BaseModel):
    a: int = 42
    b: str = "test"

class Config(BaseModel):
    subconf: SubConfig = SubConfig()
    c: float = 0.001

pydantic_config = Config()

with open("config.out", "w") as fd:
    json.dump(pydantic_config.json(), fd)
```

**Comment by Quentin-Anthony** [comment]
@mrwyattii -- Oh great! I took a look at #2395, and that solution should resolve my issue. 

Feel free to close this issue now, or I can close it once #2395 has been merged. I'm not sure how the DS team prefers to manage issues resolved by open PRs.
